Refactor Balances Logics + Abstract Balances + New Tokens module

anxolin opened the pull request:
# Summary

This PR:
- Refactor the logics for fetching balances and allowances
- Abstracts getting the balances (in preparation for bundling of transactions)
- Creates a new convenient tokens modules to encapsulate all this logic and only expose some useful hooks

> There was a bit of back and forth experimenting with different things I had to discard (I experimented not using multi-call and moving a few things to joitai, but I had no good results). I think for now makes sense to leave the multi-call REDUX implementation, and just unify logic into a module, refactor, and optimize all token fetching, give more control on the fetching logic.

## Details
WIP:
- Create module `tokens`, there we will find hooks and types for dealing with tokens
- Expose all hooks that have to do with balances form the module Facade (`index`)
- Create `useEffectiveBalances` which will return the balances and/or allowances 
- It does a refactor between the hook that fetches the balances and the allowances (99% of the code was the same). Now uses the same function.
- Exposes new information regarding the token fetching. Now you can check if there are errors, or if its fetching any particular token id needed. This information was part of the multical implementation and was not  being exposed (might be handy)

## How to review this PR
I'm sorry for the size of it. 🙏

The balance/allowance logic is rooted in the project deeply. Especially because is used by our code and code of uniswap we don't even care about.

To simplify the review, just focus on the module `tokens` and what is exported. Check `useEffectiveBalances`  that for now only delegates to `useOnchainBalancesAndAllowances`. This `useOnchainBalancesAndAllowances` delegates in the new `useOnchainBalances` and `useOnchainAllowances` implemented using a refactored function (`useOnchainErc20Amounts`).

## Not included: Future PRs
- [ ] `useCurrencyBalance` contains some handy methods that are only have refactored. I marked them as deprecated, and adapted them to delegate into the new hooks. Future PRs will improve these hooks and will expose them in the facade of the module.
- [ ] Make `useEffectiveBalances` to optionally get the allowances or not
- [ ] Implement a way to modify the balances by having some operations in the state of the app (prep for bundling of tx)
- [ ] `useAllTokenBalances` has a weird implementation, and should be re-visited
- [ ] `useCurrencyBalance` should be moved to the right module
- [ ] Delete a lot of old code from uniswap. It wasn't very pleasant to have to refactor it too!
- [ ] I will do some further refactors on some hard-to-read and maintain copy-pasted logics (like sanitizing addresses)
- [ ] Review if it's worth it to make it non-optional https://github.com/cowprotocol/cowswap/pull/2302#discussion_r1165814352

shoom3301 commented on src/custom/lib/hooks/routing/clientSideSmartOrderRouter/clientSideSmartOrderRouterMod.ts:
```diff
@@ -10,3 +10,7 @@ export function isSupportedChainId(chainId: ChainId | undefined): boolean {
   return toSupportedChainId(chainId) !== undefined
 }
+
+export function isSupportedChainIdType(chainId: number): chainId is SupportedChainId {
```
Nitpick: Can we just change the return type of `isSupportedChainId`? `boolean` => `chainId is SupportedChainId`

anxolin replied:
I tried that, but it didn't work. I took a bad assumption on why it didn't

Now i tried again, and see it works if I also change the type of the param to `chainId: number | undefined`

anxolin replied:
I'm addressing in a new PR on top of the DELETION of code to avoid messing it up

8b02fee64a18ed0295896fb0854e96571dc4bc34

shoom3301 commented on src/cow-react/modules/tokens/hooks/useOnchainBalances.ts:
```diff
@@ -0,0 +1,144 @@
+import { OnchainState } from '../types'
+
+const ERC20Interface = new Interface(ERC20ABI) as Erc20Interface
+const DEFAULT_LISTENER_OPTIONS: ListenerOptionsWithGas = { gasRequired: 185_000, blocksPerFetch: 5 }
```
Would be great to add some info about this const? Why exactly 185_000?

shoom3301 commented on src/cow-react/modules/tokens/hooks/useOnchainBalances.ts:
```diff
@@ -0,0 +1,144 @@
+
+export function useOnchainAllowances(params: OnchainAllowancesParams): OnchainAmountsResult {
+  const { account, spender } = params
+  const callParams = useMemo(() => [account, spender], [account, spender])
```
Feel free to ignore it

anxolin replied:
I thought so too, but I decided to not change the behavior of this, because I assumed it was an optimization. 
Note that if I don't use the useMemo, it will create a new array instance each time. So I thought that might trigger some additional re-renders. 

I will remove the useMemo just for simplicity, but we might need to revisit the optimizations on this hooks

shoom3301 commented on src/cow-react/modules/tokens/hooks/useOnchainBalances.ts:
```diff
@@ -0,0 +1,144 @@
+}
+
+export interface OnchainAmountsParams {
+  account?: string
```
Can we get rid of optionality here?
Would be cool to leave the responsibility of checking `account` and `tokens` falseness outside of the hook.
I mean, the hook should "think" that `account` and `tokens` always truthy

anxolin replied:
I will put it in the TODO list to review if we want to do it. For now, I kept a lot the compatibility since this is used a lot. 

I think is probably fine. This was used in some place to implement this logic:
- If the orders table is in OPEN tabs
- Then pass undefined in the account, so the hook will not return any tokens balance

I changed this so we pass an empty array of tokens instead. So probably we can do these in all the places we use the hook, but it needs to be reviewed after all refactors

shoom3301 commented on src/cow-react/modules/tokens/hooks/useOnchainBalances.ts:
```diff
@@ -0,0 +1,144 @@
+): OnchainBalancesAndAllowances {
+  const { account, spender, tokens, blocksPerFetchAllowance, blocksPerFetchBalance } = params
+
+  const { amounts: balances, isLoading: areBalancesLoading } = useOnchainBalances({
```
I would split balances and allowances, because in some cases we need only one of them.
I suggest smth like this:
```
// Updater.ts

function Updater() {
  const updateAllowancesAtom = useUpdateAtom(allowancesAtom)

  const { amounts: allowances, isLoading } = useOnchainAllowances({
    account,
    tokens,
    spender,
    blocksPerFetch: blocksPerFetchAllowance,
  })
  
  useEffect(() => {
    updateAllowancesAtom({allowances, isLoading})
  }, [allowances, isLoading])
}


// useOnchainBalancesAndAllowances.ts

function useOnchainBalancesAndAllowances() {
  const { allowances, isLoading: areAllowancesLoading } = useAtomValue(allowancesAtom)
  const { balances, isLoading: areBalancesLoading } = useAtomValue(balancesAtom)

  return useMemo(
    () => ({
      balances,
      allowances,
      isLoading: areBalancesLoading || areAllowancesLoading,
    }),
    [balances, allowances, areBalancesLoading, areAllowancesLoading]
  )
}
```

anxolin replied:
Splitting is a good idea, but I don't understand why to add joitai in the mix. 

Both are stored in REDUX already (multi-call data). I was planning on separating (the onchain hooks are already separated, I just need the "effective" ones)

shoom3301 commented on src/cow-react/modules/tokens/hooks/useOnchainBalances.ts:
```diff
@@ -0,0 +1,144 @@
+  const validatedTokenAddresses = useMemo(() => validatedTokens.map((vt) => vt.address), [validatedTokens])
+
+  // Do on-chain calls
+  const balancesCallState = useMultipleContractSingleData(
```
https://github.com/cowprotocol/cowswap/pull/2306

@anxolin @alfetopito I don't propose it, just an example.
I would like to have an opportunity to use multicall as a function, not as a hook.

alfetopito replied:
Ohh, I did that in the past.
I guess it's somewhere around in the code

anxolin commented on src/cow-react/modules/orders/hooks/useTokensFromOrders.ts:
```diff
@@ -0,0 +1,9 @@
+import { useMemo } from 'react'
+import { Order } from '@cow/modules/orders'
+
+export function useTokensFromOrders(orders: Order[]) {
```
feca466a0b24f33e7efcdded2e312d1d788e7145

anxolin commented on src/cow-react/modules/tokens/hooks/useEffectiveBalances.ts:
```diff
@@ -0,0 +1,27 @@
+ *    - Allowance: The effective balance will be the minimum between allowance and balance
+ *    - Transactions in an uncommited bundle (for gnosis safe users, it modifies the balance assuming the transaction is executed)
+ */
+export function useEffectiveBalances(params: EffectiveBalancesParams): EffectiveBalances {
```
> Why not use/rename that one directly?

In my TODO I have the continuation on this. I will add the stuff to change the balances based on the bundle

anxolin replied:
> Given the description, now I understand what this hook does.
> Still, the name is not super clear. What about useUsableBalances or useAvailableBalances ?

I will re-name to just `useBalancesAndAllowances` as opposed to the other `useOnchainBalancesAndAllowances`

anxolin replied:
I did a bunch of improvements on names 81f229e013070b8f6e535dd2a8fb4e09b2a6063f

anxolin commented on src/cow-react/modules/tokens/types.ts:
```diff
@@ -0,0 +1,7 @@
+export interface OnchainState<T> {
+  value: T
+  loading: boolean
+  syncing: boolean
```
TBH, im re-exposing the info we have from the library. My assumption is that loading will be the first load, and sync when you have a value and you are just updating. 

I will double-check and add comments

anxolin replied:
syncing is when the `block number <  your last sync` 

I added some documentation dce9f0d46348cbdcf535b58124cdba8c2695795b

anxolin commented on src/cow-react/modules/tokens/types.ts:
```diff
@@ -0,0 +1,7 @@
+  value: T
+  loading: boolean
+  syncing: boolean
+  error: boolean
+  valid: boolean
```
No exactly, 986ee406a95bcd1037105a58d365831ffc109920

anxolin commented:
@alfetopito @shoom3301 I addressed the comments, but I did it in this PR on-top of the Uniswap code deletion.

https://github.com/cowprotocol/cowswap/pull/2309

Anything else, I will address there.
